### DISCUSSION TITLE
Dynamic Text

### DISCUSSION THREAD

**Discussion by ztef** [vsg-dev/VulkanSceneGraph#891] [Q&A]
I succesfully load a font and write text using vsg::Text::create() and add to a scene, but when i try to add text dynamically there is a crash. Any Idea ?

**Comment by robertosfield** [comment]
You need to run compile traversal on newly created scene graphs so the Vulkan objects are created and any data associated with them is transferred to the GPU memory.  If you create a subgraph (including text) prior to the viewer.compile() then it'll be automatically compiled for you, but if you create it after the viewer.compile() you'll need to manually compile it using the vsg::CompileManager,

The vsgdynamicload example illustrates how this is done.

**Reply by ztef** [reply]
As I Understand in the vsgdynamicload example , do I have to compile and then merge to the viewer ? is this approach correct  ?

**Comment by ztef** [comment]
It is working now. a new subgraph Compiled and added. But added to a scene using addChild() directly. It works fine.  

But I see In the example you are using  ref_viewer->addUpdateOperation() to queue and   updateViewer(*ref_viewer, compileResult); prior the addChild, is this necessary ?

**Reply by robertosfield** [reply]
The update operation is used in the vsgdynamicload example to make sure that the main scene graph is only modified in the single threaded updated phase of the viewer's frame.  

If you allow the scene graph to be modified by any thread at any time then you'll eventually cause a crash as traversals of the scene graph can happen at the same time as that scene graph is being modified.

If you run your scene graph creation single threaded then you won't need to go through this extra step - the vsgdynamicload is design to illustrate multi-threaded loading, compiling and merging so need this extra threading protection.
